core/src/main/scala/org/apache/spark/deploy/SparkS3Util.scala

```diff
@@ -0,0 +1,299 @@
+      .reduceLeft(_ ++ _)
+  }
+
+  private def statusFromObjects(
+      bucket: String,
+      objects: util.List[S3ObjectSummary]): Iterator[LocatedFileStatus] = {
+    val blockSize: Int = conf.getInt("fs.s3n.block.size", 67108864)
+    val list: ArrayBuffer[LocatedFileStatus] = ArrayBuffer()
+    for (obj: S3ObjectSummary <- objects.asScala) {
+      if (!obj.getKey.endsWith("/")) {
+        val path = new Path(bucket + "/" + obj.getKey)
+        if (hiddenFileFilter.accept(path)) {
+          val status: FileStatus = new FileStatus(
+            obj.getSize,
+            false,
+            1,
+            blockSize,
+            obj.getLastModified.getTime,
+            path)
+          list += createLocatedFileStatus(status)
+        }
+      }
+    }
+    list.iterator
+  }
+
+  private def createLocatedFileStatus(status: FileStatus): LocatedFileStatus = {
+    val fakeLocation: Array[BlockLocation] = Array[BlockLocation]()
+    new LocatedFileStatus(status, fakeLocation)
+  }
+
+  private def listLocatedStatus(
+      s3: AmazonS3Client,
+      path: Path): RemoteIterator[LocatedFileStatus] = {
+    new RemoteIterator[LocatedFileStatus]() {
+      private final val iterator: Iterator[LocatedFileStatus] = listPrefix(s3, path)
+
+      def hasNext: Boolean = {
+        try {
+          iterator.hasNext
+        } catch {
+          case e: AmazonClientException => throw new IOException(e)
+        }
+      }
+
+      def next(): LocatedFileStatus = {
+        try {
+          iterator.next()
+        } catch {
+          case e: AmazonClientException => throw new IOException(e)
+        }
+      }
+    }
+  }
+
+  private def listStatus(s3: AmazonS3Client, paths: List[String]): Array[FileStatus] = {
+    val list: ArrayBuffer[FileStatus] = ArrayBuffer()
+    paths.foreach { path =>
+      val iterator = listLocatedStatus(s3, new Path(path))
+      while (iterator.hasNext) {
+        list += iterator.next
+      }
+    }
+    list.toArray
+  }
+
+  private def isSplitable(conf: JobConf, file: Path): Boolean = {
+    val compressionCodecs = new CompressionCodecFactory(conf)
+    val codec = compressionCodecs.getCodec(file)
+    if (codec == null) {
+      true
+    } else {
+      codec.isInstanceOf[SplittableCompressionCodec]
+    }
+  }
+
+  /**
+   * Return whether the given path is an S3 path or not.
+   */
+  def isS3Path(path: Path): Boolean = {
+    val scheme = Option(path.toUri.getScheme)
+    scheme.map(_.toUpperCase.startsWith("S3")).getOrElse(false)
+  }
+
+  /**
+   * This is based on `FileInputFormat.getSplits` method. Two key differences are:
+   *   1) Use `AmazonS3Client.listObjects` instead of `FileSystem.listStatus`.
+   *   2) Bypass data locality hints since they're irrelevant to S3 objects.
+   */
+  def getSplits(jobConf: JobConf, minSplits: Int): Array[InputSplit] = {
+    val inputDirs: Array[String] = jobConf.get(FILEINPUTFORMAT_INPUTDIR).split(",")
```

Comment:
Instead of `split` at here, can we use `getInputPaths` at here? It will properly unescape paths encoded in the string.